rust/uefi/stub/src/thin.rs

```diff
@@ -120,21 +124,67 @@ pub fn boot_linux(handle: Handle, mut system_table: SystemTable<Boot>) -> Status
         .expect("Failed to extract configuration from binary. Did you run lzbt?")
     };
 
-    let kernel_data;
-    let initrd_data;
+    let mut kernel_data;
+    let mut initrd_data;
 
     {
-        let mut file_system = system_table
-            .boot_services()
-            .get_image_file_system(handle)
-            .expect("Failed to get file system handle");
-
-        kernel_data = file_system
-            .read(&*config.kernel_filename)
-            .expect("Failed to read kernel file into memory");
-        initrd_data = file_system
-            .read(&*config.initrd_filename)
-            .expect("Failed to read initrd file into memory");
+        let filesystem_protocol_params = uefi::table::boot::OpenProtocolParams {
+            handle,
+            agent: handle,
+            controller: None
+        };
+
+        if system_table.boot_services().test_protocol::<uefi::proto::media::fs::SimpleFileSystem>(filesystem_protocol_params).is_ok() {
+            let mut file_system = system_table
+                .boot_services()
+                .get_image_file_system(handle)
+                .expect("Failed to get file system handle");
+
+            kernel_data = file_system
+                .read(&*config.kernel_filename)
+                .expect("Failed to read kernel file into memory");
+            initrd_data = file_system
+                .read(&*config.initrd_filename)
+                .expect("Failed to read initrd file into memory");
+        } else {
+            let loaded_image_protocol = system_table.boot_services().open_protocol_exclusive::<LoadedImage>(system_table.boot_services().image_handle())
+                .expect("Failed to open the loaded image protocol on the currently loaded image");
+
+            let mut base_code = system_table.boot_services().open_protocol_exclusive::<BaseCode>(loaded_image_protocol.device()).unwrap();
+
+            assert!(base_code.mode().dhcp_ack_received);
+            let dhcp_ack: &DhcpV4Packet = base_code.mode().dhcp_ack.as_ref();
+            let server_ip = dhcp_ack.bootp_si_addr;
+            let server_ip = IpAddress::new_v4(server_ip);
+
+            let kernel_filename = cstr8!("./bzImage");
```

Comment:
Ideally, we should use the embedded paths in the binary, which bring the burning question
on how do you encode those properly…

Reply:
What problems do you anticipate when loading the filenames from the binary?

Reply:
What is the path format you want to adopt for referring to filenames that can come from a non-filesystem, e.g. remote network locations?

Do we assume that we always load relative to the root of whatever "filesystem" (be it network or local)?

Reply:
> What is the path format you want to adopt for referring to filenames that can come from a non-filesystem, e.g. remote network locations?

Could we use URLs? tftp://some-server/some-file? This would also map nicely to HTTP later. The only wart is that we would have to invent a special hostname for the BOOTP server we get via DHCP.